### DISCUSSION TITLE
TGen error causes

### DISCUSSION THREAD

**Discussion by ghost** [shadow/shadow#2134] [Q&A] [answered]
I have been running large-scale Tor simulations using tornettools, which simulates clients with TGen. After some analysis, I've found that most of the errors that occur in TGen clients are "READEOF," and this seems to happen very often (up to 30% error rate with the 5 MiB stream with no padding machines enabled). However, I haven't been able to determine the source of this error. I took a brief look through the TGen source code, and there doesn't seem to be any reason why sockets are being closed early. This might make more sense if the errors were happening during a setup phase (such as SOCKS connection), but they tend to occur after a fair number of bytes have already been sent. I was hoping that someone might have some insight into this, or at least be able to point me to some resource that might be useful in figuring out the reason for these errors.

**Comment by robgjansen** [comment, marked as answer]
> I have been running large-scale Tor simulations using tornettools

I hope you're finding our tools useful :)

> up to 30% error rate with the 5 MiB stream with no padding machines enabled

Just so we're on the same page, tornettools will create two main client types: perf clients and Markov clients. I think when you stated you had "no padding machines enabled" I think you meant that you had no Markov clients running. And so I think you're asking about errors on the fixed 5MiB transfer perf clients.

30% `READEOF` errors does indeed sound bad. I don't generally get error rates that high. Something that could possibly make sense could be a timeout, since it's only happening on 5MiB downloads and only after the stream has been going for a while. I wonder if a server timeout happens, and then the server closes the connection and the EOF propagates to the client? You could look into the server TGen log files to see if anything stands out on those.

(Also, note that timeouts are more common for the Markov model clients because the probabilistic delay model can eventually lead to a timeout.)

**Reply by ghost** [reply]
> up to 30% error rate with the 5 MiB stream with no padding machines enabled

I might not have been as specific as I should have been -- I was referring to Tor circuit padding machines, meaning the Tor client is unmodified when I get the 30% error rate (with padding machines the error rate can be much higher). In these simulations, the only modification I made was to the TGen client, which I edited somewhat to print out log files that I could parse specific information out of more easily.

I am running tornettools with the standard configuration; I created my simulated networks using basically the same steps as the README except with more recent metrics data. I was under the impression that the Markov clients generated realistic traffic intended to mimic Tor users, and the usefulness of the perf clients' results depended on this. So I have 100 Markov clients and 100 perf clients at 1%.

> Something that could possibly make sense could be a timeout

I did consider the idea of a server timeout resulting in EOF being propagated to the client, but what is weird about this situation is that the server errors also seem to be read EOF. I've seen at least once simulation turn out with nothing but timeouts, but I haven't been able to replicate this recently.

Also, it might be important to note that I don't get errors with an unmodified Tor client with 50 KiB or 1 MiB downloads, but I do when the Tor client is modified. It seems that extra load has something to do with it, with one circuit padding machine in particular causing 100% error rate, but this doesn't fit any congestion control system I could think of, since packet drops would be a timeout. I thought about the possibility of Tor border nodes closing SOCKS connections, but I haven't found anything indicating this behavior.

Thanks for the insight, and any more that you might have! All of your tools have been very useful, and I'm close to concluding my research; the only problem I'm running into now is trying to understand the reasons for these errors.

**Reply by robgjansen** [reply]
> So I have 100 Markov clients and 100 perf clients at 1%.

@sporksmith @stevenengler 
- Does 100 Markov clients plus 100 perf clients seem right to you in a 1% network? This seems a bit high to me, so I'm wondering if there is legitimate overload in the network and that is what is causing the EOFs?
- Have either of you noticed read EOFs in any recent simulations?

**Reply by stevenengler** [reply]
> > So I have 100 Markov clients and 100 perf clients at 1%.
> 
> @sporksmith @stevenengler
> 
>     * Does 100 Markov clients plus 100 perf clients seem right to you in a 1% network? This seems a bit high to me, so I'm wondering if there is legitimate overload in the network and that is what is causing the EOFs?

I don't know how the number of perf/markov clients and the amount of work per perf/markov client are calculated. But if it's 100 each for a 1% network and it scales linearly, that's only 10,000 each for a 100% network which seems reasonable to me.

> * Have either of you noticed read EOFs in any recent simulations?

I haven't run any >= 1% simulations in a while. I think I always had high error rates for markov clients, but I'm not sure if those were EOFs. The error rates for perf clients were much lower.

@ethanwwitwer You're seeing this high rate of EOFs in your perf clients as well right? Can you attach some tgen perf client logs for these cases when running a simulation with an unmodified tor? Also is your simulation using any onion service traffic?

> Also, it might be important to note that I don't get errors with an unmodified Tor client with 50 KiB or 1 MiB downloads, but I do when the Tor client is modified. It seems that extra load has something to do with it, with one circuit padding machine in particular causing 100% error rate, but this doesn't fit any congestion control system I could think of, since packet drops would be a timeout. I thought about the possibility of Tor border nodes closing SOCKS connections, but I haven't found anything indicating this behavior.

Do any of your tor logs show a message at notice level containing "We're low on memory"?

> I did consider the idea of a server timeout resulting in EOF being propagated to the client, but what is weird about this situation is that the server errors also seem to be read EOF. I've seen at least once simulation turn out with nothing but timeouts, but I haven't been able to replicate this recently.

Are the EOFs seen on perf clients only? Or also the servers?

**Reply by ghost** [reply]
I haven't actually looked at error rates for Markov clients as of yet, just perf clients. I've mostly been reviewing the error rates in aggregate and noticed that a large percentage of them were read EOFs, but after reviewing the logs further it does seem that there are mostly just timeouts in the unmodified Tor simulations. It is the simulations with circuit padding machines that primarily produce read EOF errors. In these simulations, the EOFs occur in the clients and servers. As far as configuration, I'm not simulating onion service traffic in any of the simulations, and I am using several independently-developed padding machines in the modified simulations, so that makes it seem less likely that implementation issues are to blame.

It does appear that memory is an issue. tornettools produces a file _free.log_, which appears to just contain the output of the _free_ command called repeatedly. After parsing these files, it looks like the simulations with the greatest error rates exhaust almost all of the system's memory (minimum value of available memory is less than 1 GB), whereas the unmodified simulations leave over 120 GB remaining at peak, and other simulations are somewhere in-between. I was aware of this before starting this discussion thread, but I couldn't figure out, if it was related, the exact mechanism by which it resulted in the EOFs. In the Tor logs, I only found this:

> Jan 01 00:00:04.000 [notice] Based on detected system memory, MaxMemInQueues is set to 8192 MB. You can override this by setting MaxMemInQueues by hand.

This doesn't seem very significant to me, since there is a lot more memory at play here than 8 MiB per process.

I have attached a few of the log files for an unmodified simulation and a few for the modified one with the highest error rate.

[control-1-stdout.txt](https://github.com/shadow/shadow/files/8683967/control-1-stdout.txt)
[control-2-stdout.txt](https://github.com/shadow/shadow/files/8683959/control-2-stdout.txt)
[control-3-stdout.txt](https://github.com/shadow/shadow/files/8683961/control-3-stdout.txt)
[modified-1-stdout.txt](https://github.com/shadow/shadow/files/8683963/modified-1-stdout.txt)
[modified-2-stdout.txt](https://github.com/shadow/shadow/files/8683964/modified-2-stdout.txt)
[modified-3-stdout.txt](https://github.com/shadow/shadow/files/8683965/modified-3-stdout.txt)

**Reply by stevenengler** [reply]
> It does appear that memory is an issue. tornettools produces a file _free.log_, which appears to just contain the output of the _free_ command called repeatedly. After parsing these files, it looks like the simulations with the greatest error rates exhaust almost all of the system's memory (minimum value of available memory is less than 1 GB), whereas the unmodified simulations leave over 120 GB remaining at peak, and other simulations are somewhere in-between. I was aware of this before starting this discussion thread, but I couldn't figure out, if it was related, the exact mechanism by which it resulted in the EOFs. In the Tor logs, I only found this:
> 
> > Jan 01 00:00:04.000 [notice] Based on detected system memory, MaxMemInQueues is set to 8192 MB. You can override this by setting MaxMemInQueues by hand.
> 
> This doesn't seem very significant to me, since there is a lot more memory at play here than 8 MiB per process.

Tor monitors its memory usage and if it becomes near this limit, it will start destroying circuits. This would probably show up as an EOF for tgen streams. But if you aren't seeing any "We're low on memory" log messages, then this doesn't appear to be happening.

When the tornettools simulation finishes, what does the tonettools log say the returned code was? (What's the code in the log line "Done simulating; shadow returned code '_'")?

**Reply by stevenengler** [reply]
Can you also run one of your memory-intensive simulations and check for oom-killed processes immediately after with something like `grep -i 'killed process' /var/log/syslog` (might need a different command depending on your distribution)?

**Reply by ghost** [reply]
Shadow returns code 0 for all of these 1% simulations, even the ones which exhaust all of the system's memory, but if I go above about 1.5% it starts returning error codes. I checked again to confirm that none of the Tor logs have any messages about lack of memory. Unfortunately, I don't have read access to _/var/log/syslog_ on this machine, but it might be somewhat telling that the simulation that uses the most memory always outputs a similar minimum value of available memory in _free.log_, perhaps some sort of threshold below which processes begin to be killed?

**Reply by stevenengler** [reply]
That's what I'm thinking as well. I also don't think Shadow is correctly handling processes that being killed. I made an issue for that in #2151.

**Reply by ghost** [reply]
For now, I'll try running some scaled-down simulations to see if that changes much about the errors. It seems a bit weird that increasing the scale further results in Shadow returning an error code, maybe because memory issues are happening earlier, but there definitely seems to be a correlation.
